REFERENCE.md

```diff
@@ -0,0 +1,277 @@
+
+
+#### Rule Dimensions
+
+**Rule Dimensions**: defines the allowed list of Data Quality rule dimensions that a Data Quality `Rule` can define in the corresponding `dimension` field.
+
+```yaml
+rule_dimensions:
+  - consistency
+  - correctness
+  - duplication
+  - completeness
+  - conformance
+  - integrity
+```
+This list can be customized as is required. CloudDQ throws an error when parsing the YAML if it finds a `rule` with a `dimension` value that is not one of the allowed values.
+
+#### Rules
+
+**Rules**: Defines reusable logic for data quality validation.
+
+```yaml
+rules:
+  NOT_NULL_SIMPLE:
+    rule_type: NOT_NULL
+    dimension: completeness
+
+  NOT_BLANK:
+    rule_type: NOT_BLANK
+    dimension: completeness
+
+  REGEX_VALID_EMAIL:
+    rule_type: REGEX
+    dimension: conformance
+    params:
+      pattern: |-
+        ^[^@]+[@]{1}[^@]+$
+
+  CUSTOM_SQL_LENGTH_LE_30:
+    rule_type: CUSTOM_SQL_EXPR
+    dimension: correctness
+    params:
+      custom_sql_expr: |-
+        LENGTH( $column ) <= 30
+
+  VALUE_ZERO_OR_POSITIVE:
+    rule_type: CUSTOM_SQL_EXPR
+    dimension: correctness
+    params:
+      custom_sql_expr: |-
+        $column >= 0
+
+  CORRECT_CURRENCY_CODE:
+    rule_type: CUSTOM_SQL_EXPR
+    dimension: conformance
+    params:
+      custom_sql_expr: |-
+        $column in (select distinct currency_code from `<reference_dataset-id>.currency_codes`)
+
+  FOREIGN_KEY_VALID:
+    rule_type: CUSTOM_SQL_EXPR
+    dimension: consistency
+    params:
+      custom_sql_expr: |-
+        $column in (select distinct foreign_key from `<dataset-id>.customer_id`)
+
+  NO_DUPLICATES_IN_COLUMN_GROUPS:
+    rule_type: CUSTOM_SQL_STATEMENT
+    dimension: duplication
+    params:
+      custom_sql_arguments:
+        - column_names
+      custom_sql_statement: |-
+        select a.*
+        from data a
+        inner join (
+          select
+            $column_names
+          from data
+          group by $column_names
+          having count(*) > 1
+        ) duplicates
+        using ($column_names)
+```
+
+`CUSTOM_SQL_EXPR` will apply the SQL expression logic defined in `custom_sql_expr` to every record and flag any row where `custom_sql_expr` evaluated to False as a failure. `CUSTOM_SQL_EXPR` rules are meant for record-level validation and will report success/failure counts at the record level. 
+
+`CUSTOM_SQL_STATEMENT` is meant for set-level validation and instead accepts a complete SQL statement in the field `custom_sql_statement`. If the SQL statement returns 0 rows, `CloudDQ` considers the rule a success. If the SQL statement returns any row, `CloudDQ` will report the rule as a failure and include the returned row counts in the validation summary.
+
+> TODO Note Andrew: Suggest we bring up more clearly the distinction between row- and set-level validations, following with how custom_sql_expr/statement (and built-in rules) support them
```

Comment:
pls remove all TODO tags across the documents

Reply:
ie this shouldn't be publicly visible

Reply:
OK